examples/llama/pippy_llama.py

```diff
@@ -0,0 +1,109 @@
+# Minimum effort to run this example:
+# $ torchrun --nproc-per-node 2 pippy_llama.py
+
+import argparse
+import os
+
+import torch
+import torch.distributed as dist
+
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from pippy.IR import Pipe, PipeSplitWrapper, annotate_split_points
+from pippy.PipelineStage import PipelineStage
+
+
+def add_split_points(llama, nranks):
+    layers_per_rank = (llama.config.num_hidden_layers + nranks - 1) // nranks
+    print(f"layers_per_rank = {layers_per_rank}")
+    for i in range(1, nranks):
+        annotate_split_points(
+            llama,
+            {f'model.layers.{i * layers_per_rank}': PipeSplitWrapper.SplitPoint.BEGINNING},
+        )
+
+
+def get_number_of_params(model):
+    return sum(p.numel() for p in model.parameters() if p.requires_grad)
+
+
+def run(args):
+    # Create a blank model
+    llama = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-7b-chat-hf", low_cpu_mem_usage=True)
+    tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-chat-hf")
+
+    prompts = (
+        "How do you", "I like to", "Can I help", "You have to",
+        "The weather is", "I have a", "What is your", "You are a",
+    )  # bs = 8
+    tokenizer.pad_token = tokenizer.eos_token
+    inputs = tokenizer(prompts, return_tensors="pt", padding=True).to(args.device)
+
+    # Move model to `device` and set to evaluation
+    llama.to(args.device)
+    llama.eval()
+    print(llama)
+
+    # Annotate split points
+    add_split_points(llama, args.world_size)
+
+    # Create a pipeline stage from the model
+    llama_pipe = Pipe.from_tracing(
+        llama,
+        num_chunks=args.world_size,
+        example_args=(inputs['input_ids'],),
+    )
+
+    assert len(list(llama_pipe.split_gm.children())) == args.world_size
+    if args.rank == 0:
+        for i, sm in enumerate(llama_pipe.split_gm.children()):
+            print(f"Pipeline stage {i} {get_number_of_params(sm) // 10 ** 6}M params")
+
+    # Create schedule runtime
+    stage = PipelineStage(
+        llama_pipe,
+        args.rank,
+        device=args.device,
+    )
+
+    # Run
+    output = None
+    if args.rank == 0:
+        stage(inputs['input_ids'])
+    elif args.rank == args.world_size - 1:
+        output = stage()
+    else:
+        stage()
+
+    if output is not None:
```

Comment:
@kwen2501 I wonder if we are using HF generate function under the hood/ we have our own implementation/ if we need any special treatment to work with HF generate function?

Reply:
No, we are not using HF's generate function here. Just a forward pass of the model.